### ISSUE TITLE
`BindNever`, `JsonIgnore`, and `BindRequired`

### ISSUE THREAD

**Issue by Biarity** [json-api-dotnet/JsonApiDotNetCore#233]
When using the entity model as the resource model, it is useful to be able to use the `BindNever`, `JsonIgnore`, and `BindRequired` attributes. This way you can easily specify which properties should be hidden from the client and how. As far as I know, you can't do this with the current implementation (these attributes simply don't work).

**Comment by jaredcnance** [comment]
Attributes are hidden by default. Unless you explicitly include the `Attr` attribute, members will not be available in the public API. In our current implementation, we return a `400` if a request is made with a member that is not exposed in the public API. Is there a scenario or use case this does not cover?

**Comment by Biarity** [comment]
Yes, that kind of covers functionality for `BindNever` and `JsonIgnore` combined. But this does not allow quickly creating read-only properties (ie. using `BindNever` only - this is equivalent to using `isImmutable` being true also for POST requests) or making properties required (ie. using `BindRequired` - as far as I'm aware, you'll get an error even if you try other validation attributes). 

Another benefit that would come from somehow supporting these attributes is making it easier to adopt JsonApiDotNetCore in existing projects where models already exist with these standard attributes on them.

**Comment by jaredcnance** [comment]
Ah, I see. So, there are 2 requirements currently not met:

1. Allow read-only attributes which cannot be set on `POST`
2. Improve portability of pre-existing models by supporting Newtonsoft attributes and ASP.Net model binding attributes

By implementing 2 we also implement 1. I think this will take a little bit of research to do properly. When I have time I'll spec out the requirements for the implementation in this issue.

**Comment by bart-degreed** [comment]
I've had some discussion with @maurei on this today. We found that supporting the Newtonsoft.Json attributes on resource properties is not straightforward to implement, because we do not directly serialize the resource class itself but an intermediate `ResourceObject` type, in which we copy (a subset of) property values based on sparse fieldsets (see #796 for details).

Because the runtime-type of the property value gets preserved, the serializer does pick up Newtonsoft attributes on the property type, just not on the property itself. To add support for that, we'd need to either duplicate handling logic of the various Newtonsoft attributes or split/join serialized strings for the individual properties (which messes up things like indents, line breaks etc).

Suggestions to implement this cleanly are welcome.

**Comment by bart-degreed** [comment]
As mentioned earlier, we cannot propagate all custom attributes on resource properties. Our recent switch to System.Text.Json does not change that.

However:
- We do provide `AttrCapabilities`, which enables devs to specify setting a value on POST and/or PATCH is not allowed.
- We have working ASP.NET model state validation, so you can put `[Required]` on properties. This works correctly when combined with partial patch. Note this is turned off by default.

I believe these address the requested functionality, so I'm closing this. It does not enable reusing existing models with Newtonsoft or System.Text.Json annotations, but as explained earlier I don't think we can solve that.
